### PR TITLE
Table: Datepicker, AutoComplete, Select Cell Editor

### PR DESCRIPTION
**Pull request by ElishaSamPeterPrabhu**
## Description

- Updated dropdown, autocomplete to accept badge and links args and  datepicker of cell to handle out of focus

References 

#2310 , #2311 , #2313 
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

https://deploy-preview-2387--moduswebcomponents.netlify.app/

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by cjwinsor** [comment]
@ElishaSamPeterPrabhu Some observations:
Storybook:
- Autocomplete should have all First Names from `Names` array
- Autocomplete for the last row will have its items require the user to scroll down in the container to see, p
- CreatedAt should be of type Date so we can show that too
- Email column's Autocomplete options should contain all emails

Component:
- Autocomplete should set the value
- Autocomplete should show items on focus
- Autocomplete, you can't use arrows or tab to navigate the items in the list, table handling takes over. (arrows work for number)
- Autocomplete increases the row height
- Date should set value
- Date column and editor formatting should match
- Date doesn't lose focus when you click away
- Date Calendar is inside the overflow container, any way to pop it out? Last item you can look at

**Comment by cjwinsor** [comment]
@ElishaSamPeterPrabhu I went over these changes again and this is what I think is needed before we can merge:

- Remove the `overflow: auto` from `.table-container` in modus-table.scss.  Instead modus-table.tsx should set `overflow: auto` only when a `maxHeight` is provided. This way, if no `maxHeight` is defined then the datepicker and autocomplete won't be hidden inside the overflow container.  It doesn't fix for when `maxHeight` is defined, but at gives a workaround until we can address this.
-  Switch from using "dropdown" for the Select component. Use "select" instead.  Update all code to match. This is to avoid confusion with our drop down component we have already which is different.  
- Don't reuse `ModusTableCellDropdownEditorArgs` for Autocomplete, create a new type for it.  
- `ModusTableCellDropdownEditorArgs` should have `optionsDisplayProp` as well, in addition to `options`.
- The `ModusTableCellAutocompleteEditorArgs` which you'll have created above, should have the following additional properties (beyond just options): `noResultsFoundText`, `noResultsFoundSubtext`, `showNoResultsFoundMessage`, `showOptionsOnFocus`. 
- `ModusTableCellAutocompleteEditorArgs.options` should be of type `ModusAutocompleteOption `
- For autocomplete, it seems if i click on the cell in the whitepace, then click on another cell, it loses focus correctly, but if i click on the text then another cell it doesn't loose focus.  Might be related to the example being a link, I didn't check with a non link autocomplete column
- I think we need to make sure `ModusTableCellEditorType` in storybook and code shows and supports all the different cell editor types, it currently only shows `type ModusTableCellEditorType = typeof CELL_EDIT_TYPE_DROPDOWN;`

**Comment by ElishaSamPeterPrabhu** [comment]
> @ElishaSamPeterPrabhu I went over these changes again and this is what I think is needed before we can merge:
> 
> * Remove the `overflow: auto` from `.table-container` in modus-table.scss.  Instead modus-table.tsx should set `overflow: auto` only when a `maxHeight` is provided. This way, if no `maxHeight` is defined then the datepicker and autocomplete won't be hidden inside the overflow container.  It doesn't fix for when `maxHeight` is defined, but at gives a workaround until we can address this.
> * Switch from using "dropdown" for the Select component. Use "select" instead.  Update all code to match. This is to avoid confusion with our drop down component we have already which is different.
> * Don't reuse `ModusTableCellDropdownEditorArgs` for Autocomplete, create a new type for it.
> * `ModusTableCellDropdownEditorArgs` should have `optionsDisplayProp` as well, in addition to `options`.
> * The `ModusTableCellAutocompleteEditorArgs` which you'll have created above, should have the following additional properties (beyond just options): `noResultsFoundText`, `noResultsFoundSubtext`, `showNoResultsFoundMessage`, `showOptionsOnFocus`.
> * `ModusTableCellAutocompleteEditorArgs.options` should be of type `ModusAutocompleteOption `
> * For autocomplete, it seems if i click on the cell in the whitepace, then click on another cell, it loses focus correctly, but if i click on the text then another cell it doesn't loose focus.  Might be related to the example being a link, I didn't check with a non link autocomplete column
> * I think we need to make sure `ModusTableCellEditorType` in storybook and code shows and supports all the different cell editor types, it currently only shows `type ModusTableCellEditorType = typeof CELL_EDIT_TYPE_DROPDOWN;`

Have worked on all of the above suggestions
Regarding the autocomplete focus behavior its seems to be working fine :

https://github.com/trimble-oss/modus-web-components/assets/122255626/05bddb9d-d570-4f9e-890f-ac779c4adf27

**Review by cjwinsor — CHANGES_REQUESTED** [review]

File: stencil-workspace/src/components/modus-table/modus-table.tsx [hunk 1]
```diff
@@ -86,6 +86,10 @@ import { createGuid } from '../../utils/utils';
 export class ModusTable {
   @Element() element: HTMLElement;
 
+  @Prop() editableCellId: string[] = [];
+  @Watch('editableCellId') onEditableCellIdsChange(newVal: string[]) {
+    console.log('editableCellIds', newVal);
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/modus-table.tsx:91]: Oops committed it by mistake, will remove it

File: stencil-workspace/src/components/modus-table/modus-table.tsx [hunk 2]
```diff
@@ -373,6 +377,7 @@ export class ModusTable {
     this.setTableState(initialTableState);
     this.onRowsExpandableChange(this.rowsExpandable);
     this.initializeTable();
+    console.log('editableCellId232s', this.editableCellId);
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/modus-table.tsx:380]: Removed

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 3]
```diff
@@ -121,45 +143,72 @@ export class ModusTableCellEditor {
...
+          }}></modus-date-input>
       </modus-date-picker>
     );
   }
 
   renderAutocompleteInput(): JSX.Element[] {
-    const options = ((this.args as ModusTableCellDropdownEditorArgs)?.options || []) as ModusAutocompleteOption[] | string[];
+    let options, selectedOption;
+    const args = this.args as ModusTableCellAutocompleteEditorArgs;
+    // const { noResultsFoundText, noResultsFoundSubtext, showNoResultsFoundMessage, showOptionsOnFocus } = args;
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:163]: Removed

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 4]
```diff
@@ -121,45 +143,72 @@ export class ModusTableCellEditor {
...
-        }}
-        // value={selectedOption}
-        // onKeyDown={(e) => e.stopPropagation()}
-      ></modus-autocomplete>
+      <div class="autocomplete-container">
+        <modus-autocomplete
+          {...this.getDefaultProps('Autocomplete input')}
+          include-search-icon="false"
+          size="medium"
+          //onClick={(e: MouseEvent) => e.stopPropagation()}
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:180]: Removed

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 5]
```diff
@@ -25,28 +26,40 @@ import { ModusAutocompleteOption } from '../../../../modus-autocomplete/modus-au
...
     if (this.inputElement['focusInput']) this.inputElement['focusInput']();
+
+    this.outsideClickListener = (event) => {
+      if (!this.inputElement.contains(event.target)) {
+        this.handleBlur();
+      }
+    };
+    if (this.type == 'date') document.addEventListener('click', this.outsideClickListener);
+  }
+  disconnectedCallback(): void {
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:53]: Replaced the callback logic with listen

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 6]
```diff
@@ -25,28 +26,40 @@ import { ModusAutocompleteOption } from '../../../../modus-autocomplete/modus-au
...
+  @Prop() dataType: string;
+  @Prop() value: any;
   @Prop() type: string;
   @Prop() valueChange: (newValue: string) => void;
   @Prop() keyDown: (e: KeyboardEvent, newValue: string) => void;
 
-  private editedValue: string;
+  private editedValue: any;
   private inputElement: HTMLElement;
+  private outsideClickListener: (event: any) => void;
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:37]: Yes changed

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 7]
```diff
@@ -25,28 +26,40 @@ import { ModusAutocompleteOption } from '../../../../modus-autocomplete/modus-au
...
   @Prop() args: ModusTableCellEditorArgs;
-  @Prop() value: string;
+  @Prop() dataType: string;
+  @Prop() value: any;
   @Prop() type: string;
   @Prop() valueChange: (newValue: string) => void;
   @Prop() keyDown: (e: KeyboardEvent, newValue: string) => void;
 
-  private editedValue: string;
+  private editedValue: any;
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:35]: Replaced with unknown

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 8]
```diff
@@ -25,28 +26,40 @@ import { ModusAutocompleteOption } from '../../../../modus-autocomplete/modus-au
 })
 export class ModusTableCellEditor {
   @Prop() args: ModusTableCellEditorArgs;
-  @Prop() value: string;
+  @Prop() dataType: string;
+  @Prop() value: any;
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:30]: Replaced with unknown

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 9]
```diff
@@ -121,45 +143,72 @@ export class ModusTableCellEditor {
...
   renderAutocompleteInput(): JSX.Element[] {
-    const options = ((this.args as ModusTableCellDropdownEditorArgs)?.options || []) as ModusAutocompleteOption[] | string[];
+    let options, selectedOption;
+    const args = this.args as ModusTableCellAutocompleteEditorArgs;
+    // const { noResultsFoundText, noResultsFoundSubtext, showNoResultsFoundMessage, showOptionsOnFocus } = args;
+    if (this.dataType === 'badge') {
+      options = args?.options.map((option: any) => option.text) as ModusAutocompleteOption[] | string[];
+      selectedOption = this.value['text'];
+    } else if (this.dataType === 'link') {
+      options = args?.options.map((option: any) => option.display) as ModusAutocompleteOption[] | string[];
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:168]: Replaced with unknown

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 10]
```diff
@@ -121,45 +143,72 @@ export class ModusTableCellEditor {
...
+          {...this.getDefaultProps('Autocomplete input')}
+          include-search-icon="false"
+          size="medium"
+          //onClick={(e: MouseEvent) => e.stopPropagation()}
+          options={options}
+          onBlur={this.handleBlur}
+          onKeyDown={(e) => e.stopPropagation()}
+          onOptionSelected={(e: CustomEvent<string>) => {
+            if (this.dataType === 'badge') {
+              args?.options.map((option: any) => {
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:186]: Replaced with unknown

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 11]
```diff
@@ -121,45 +143,72 @@ export class ModusTableCellEditor {
...
+                if (option.display == e.detail) {
+                  this.editedValue = option;
+                }
+              });
+            } else {
+              this.editedValue = e.detail;
+            }
+          }}
+          value={selectedOption}
+          // onKeyDown={(e) => e.stopPropagation()}
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:202]: Removed

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 12]
```diff
@@ -88,10 +101,12 @@ export class ModusTableCellEditor {
...
 
-  renderDropdownInput(): JSX.Element[] {
+  renderSelectInput(): JSX.Element[] {
     const valueKey = 'display';
-    const options = (this.args as ModusTableCellDropdownEditorArgs)?.options || [];
-    const selectedOption = options.find((option) => option[valueKey] === this.value);
+    const args = this.args as ModusTableCellSelectEditorArgs;
+    const options = args?.options || [];
+    const optionsDisplayProp = args?.optionsDisplayProp || valueKey;
+    const selectedOption = options.find((option) => option[optionsDisplayProp] === this.value) as any;
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:109]: Replaced with unknown

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 13]
```diff
@@ -105,13 +120,20 @@ export class ModusTableCellEditor {
...
-          options-display-prop="display"
+          options-display-prop={optionsDisplayProp}
           size="large"
           options={options}
           onInputBlur={this.handleBlur}
           onKeyDown={(e) => handleEnter(e, this.handleKeyDown)}
           onValueChange={(e: CustomEvent<unknown>) => {
-            this.editedValue = e.detail[valueKey];
+            if (this.dataType === 'badge') {
+              const { display, ...restProps } = e.detail as any;
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:130]: Replaced with unknown

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 14]
```diff
@@ -121,45 +143,72 @@ export class ModusTableCellEditor {
...
     );
   }
 
   renderAutocompleteInput(): JSX.Element[] {
-    const options = ((this.args as ModusTableCellDropdownEditorArgs)?.options || []) as ModusAutocompleteOption[] | string[];
+    let options, selectedOption;
+    const args = this.args as ModusTableCellAutocompleteEditorArgs;
+    // const { noResultsFoundText, noResultsFoundSubtext, showNoResultsFoundMessage, showOptionsOnFocus } = args;
+    if (this.dataType === 'badge') {
+      options = args?.options.map((option: any) => option.text) as ModusAutocompleteOption[] | string[];
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:165]: Replaced with unknown

File: stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx [hunk 15]
```diff
@@ -25,28 +26,40 @@ import { ModusAutocompleteOption } from '../../../../modus-autocomplete/modus-au
...
+  private outsideClickListener: (event: any) => void;
 
   connectedCallback(): void {
     this.editedValue = this.value;
   }
 
   componentDidLoad(): void {
     if (this.inputElement['focusInput']) this.inputElement['focusInput']();
+
+    this.outsideClickListener = (event) => {
```
[the remark this answers is not in the document]
Comment (ElishaSamPeterPrabhu) [reply stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell-editor/modus-table-cell-editor.tsx:46]: Replaced with the above logic

**Comment by ElishaSamPeterPrabhu** [comment]
For the autocomplete focus issue: 
Added `this.cellfocus()` when a link is clicked instead of the white space this handles the out of focus issue as expected
PS: The link issue of autocomplete was present in the badge type as well when clicking on the badge directly instead of the whitespace, so added the same logic to badge as well.

**Review by cjwinsor — APPROVED** [review]

**Pull request merged**